src/shogun/features/CombinedDotFeatures.cpp

```diff
@@ -309,32 +315,35 @@ int32_t CCombinedDotFeatures::get_nnz_features_for_vector(int32_t num) const
 	return result;
 }
 
-SGVector<float64_t> CCombinedDotFeatures::get_subfeature_weights()
+SGVector<float64_t> CCombinedDotFeatures::get_subfeature_weights() const
 {
 	int32_t num_weights = get_num_feature_obj();
 	ASSERT(num_weights > 0)
 
 	float64_t* weights=SG_MALLOC(float64_t, num_weights);
+	std::copy(feature_weights.begin(), feature_weights.end(), weights);
 
-	for (index_t f_idx=0; f_idx<num_weights; f_idx++)
-	{
-		CDotFeatures* f = get_feature_obj(f_idx);
-		weights[f_idx] = f->get_combined_feature_weight();
-		SG_UNREF(f);
-	}
 	return SGVector<float64_t>(weights,num_weights);
 }
 
 void CCombinedDotFeatures::set_subfeature_weights(SGVector<float64_t> weights)
```

Comment:
right now the weights are passed by value, maybe pass by const ref? otherwise you are calling constructor right?

Reply:
sorry this is unrelated to your PR, just noticed it

Reply:
SGVector is passed by value, but the copy constructor only copies the pointer to the data. I think constant reference would be better though.

Reply:
yup!

Reply:
there are probably lots of other things like this in shogun so can probably leave it for another PR

Reply:
if you copy inside, then a const ref would be better. We use the copy by value for SG* for the reference counting magic